[MoE] Remove a sync point in get_occurrence_numbers

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* __->__ #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* #165482
* #165101
* #164181
* #164164

The number of unique indices are known ahead of time, i.e. number of experts. Thus no need to:
```
   unique_vals, inverse = torch.unique(topk_indices, return_inverse=True)
   n_unique = len(unique_vals)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/nn/parallel/moe.py:
```diff
@@ -16,7 +16,7 @@ def _get_occurrence_numbers(topk_indices: torch.Tensor) -> torch.Tensor:
+    n_unique = n_experts
     n_elements = len(topk_indices)
 
     # Create a matrix where each row corresponds to a unique value
```
Why do we need torch.arange as an accessory below, why can't we just do [inverse, :] ?

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update on "[MoE] Remove a sync point in get_occurrence_numbers"


The number of unique indices are known ahead of time, i.e. number of experts. Thus no need to:
```
   unique_vals, inverse = torch.unique(topk_indices, return_inverse=True)
   n_unique = len(unique_vals)
```

[ghstack-poisoned]
